client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -444,44 +347,112 @@ private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
-        final HostTuple<Set<HttpHost>> hostTuple = this.hostTuple;
-        Collection<HttpHost> nextHosts = Collections.emptySet();
-        do {
-            Set<HttpHost> filteredHosts = new HashSet<>(hostTuple.hosts);
-            for (Map.Entry<HttpHost, DeadHostState> entry : blacklist.entrySet()) {
-                if (System.nanoTime() - entry.getValue().getDeadUntilNanos() < 0) {
-                    filteredHosts.remove(entry.getKey());
+    private NodeTuple<Iterator<HttpHost>> nextNode(NodeSelector nodeSelector) throws IOException {
+        NodeTuple<List<Node>> nodeTuple = this.nodeTuple;
+        List<HttpHost> hosts = selectHosts(nodeTuple, blacklist, lastNodeIndex, System.nanoTime(), nodeSelector);
+        return new NodeTuple<>(hosts.iterator(), nodeTuple.authCache);
+    }
+
+    static List<HttpHost> selectHosts(NodeTuple<List<Node>> nodeTuple,
+            Map<HttpHost, DeadHostState> blacklist, AtomicInteger lastNodeIndex,
+            long now, NodeSelector nodeSelector) throws IOException {
+        class DeadNodeAndRevival {
+            final Node node;
+            final long nanosUntilRevival;
+
+            DeadNodeAndRevival(Node node, long nanosUntilRevival) {
+                this.node = node;
+                this.nanosUntilRevival = nanosUntilRevival;
+            }
+
+            @Override
+            public String toString() {
+                return node.toString();
+            }
+        }
+
+        /*
+         * Sort the nodes into living and dead lists.
+         */
+        List<Node> livingNodes = new ArrayList<>(nodeTuple.nodes.size() - blacklist.size());
+        List<DeadNodeAndRevival> deadNodes = new ArrayList<>(blacklist.size());
+        for (Node node : nodeTuple.nodes) {
+            DeadHostState deadness = blacklist.get(node.getHost());
+            if (deadness == null) {
+                livingNodes.add(node);
+                continue;
+            }
+            long nanosUntilRevival = now - deadness.getDeadUntilNanos();
+            if (nanosUntilRevival > 0) {
+                livingNodes.add(node);
+                continue;
+            }
+            deadNodes.add(new DeadNodeAndRevival(node, nanosUntilRevival));
+        }
+
+        if (false == livingNodes.isEmpty()) {
+            /*
+             * Normal state: there is at least one living node. Rotate the
+             * list so subsequent requests to will see prefer the nodes in
+             * a different order then run them through the NodeSelector so
+             * it can have its say in which nodes are ok and their ordering.
+             * If the selector is ok with any over the living nodes then use
+             * them for the request.
+             */
+            Collections.rotate(livingNodes, lastNodeIndex.getAndIncrement());
+            List<Node> selectedLivingNodes = nodeSelector.select(livingNodes);
+            if (false == selectedLivingNodes.isEmpty()) {
```

Comment:
I am tempted to just throw exception here if the selector rejected all of our alive nodes. Now that the selector has more power given that it knows what all the possible nodes are, it should try not to reject all of them? Otherwise the selector ends up forcing us to resurrect nodes which is not that great? What do you think?

Reply:
I think we should revive nodes if the selector rejects all living nodes. I expect folks that use the sniffer will also want to use the `NOT_MASTER_ONLY` selector. They will really want to revive the other nodes if they fail.

Reply:
I think you are right, I will try to find out what other language clients do in this situation just to make sure.

Reply:
Maybe at this point the revival process should also make up a sorted list with all of the dead nodes and pass that one to the selector, although we will try to revive only the first one in the list? Not too sure though, just a wild idea.

Reply:
It does that!

Reply:
I pushed some more code a few minutes ago, it might contain that change. I'm fairly sure it contains some javadoc explaining that change though.

Reply:
nice